#!/usr/bin/env python

PACKAGE = "espros_cam660"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()


lens_type_enum = gen.enum([ gen.const("WideField", int_t, 0, "WideField"),
                             gen.const("StandardField",    int_t, 1, "StandardField"),
                             gen.const("NarrowField", int_t, 2, "NarrowField")],
                             "An enum to set lens type")


image_type_enum = gen.enum([ gen.const("Grayscale", int_t, 0, "Grayscale"),
                             gen.const("Distance",    int_t, 1, "Distance"),
                             gen.const("DistranceAmplitude", int_t, 2, "Distance&Amplitude")
                             #gen.const("DCS", int_t, 3, "DCS")
                             ],
                             "An enum to set image type")

hdr_mode_enum  =  gen.enum([ gen.const("HDR_off",      int_t, 0, "HDR off"),
                             gen.const("HDR_spatial",  int_t, 1, "HDR spatial"),
                             gen.const("HDR_temporal", int_t, 2, "HDR temporal")],
                             "An enum to set HDR mode")


frequence_modulation_enum = gen.enum([ gen.const("24MHz", int_t, 1, "24MHz"),
                                       gen.const("12MHz", int_t, 0, "12MHz")],
                                       "An enum to set modulation frequency")


channel_enum = gen.enum([ gen.const("0", int_t, 0, "0"),
                          gen.const("1", int_t, 1, "1"),
                          gen.const("2", int_t, 2, "2"),
                          gen.const("3", int_t, 3, "3"),
                          gen.const("4", int_t, 4, "4"),
                          gen.const("5", int_t, 5, "5"),
                          gen.const("6", int_t, 6, "6"),
                          gen.const("7", int_t, 7, "7"),
                          gen.const("8", int_t, 8, "8"),
                          gen.const("9", int_t, 9, "9"),
                          gen.const("10", int_t, 10, "10"),
                          gen.const("11", int_t, 11, "11"),
                          gen.const("12", int_t, 12, "12"),
                          gen.const("13", int_t, 13, "13"),
                          gen.const("14", int_t, 14, "14"),
                          gen.const("15", int_t, 15, "15")],
                          "An enum to set channel index")


######  Name                    Type   Reconfiguration level  Description     Default Min  Max
### Acquisition Control
gen.add("lens_data",              str_t, 0, "Lens data file name",  "lens_wf")
gen.add("lens_center_offset_x",   int_t, 0, "Lens optical center offset X direction [pix]",  0, -50, 50)
gen.add("lens_center_offset_y",   int_t, 0, "Lens optical center offset Y direction [pix]",  0, -50, 50)
gen.add("frame_rate",          double_t, 0, "Image acquisition frame rate frame/sec",  20, 0, 30)
gen.add("image_type",             int_t, 0, "Image type", 1, 0, 2, edit_method=image_type_enum)
gen.add("start_stream",          bool_t, 0, "Start / Stop image acquisition",  False)
gen.add("trigger_single_shot",   bool_t, 0, "Trigger single image shot",  False)
gen.add("hdr_mode",               int_t, 0, "HDR mode", 0, 0, 2, edit_method=hdr_mode_enum)
gen.add("integration_time_tof_1", int_t, 0, "Integration time TOF1 [uS]",          50, 0, 4000)
gen.add("integration_time_tof_2", int_t, 0, "Integration time TOF2 [uS]",         400, 0, 1000)
gen.add("integration_time_tof_3", int_t, 0, "Integration time TOF3 [uS]",           0, 0, 2000)
gen.add("integration_time_gray",  int_t, 0, "Integration time Grayscale [uS]", 3000, 0, 100000)
gen.add("min_amplitude",          int_t, 0, "threshold minAmplitude LSB",       100, 0, 100000)
gen.add("frequency_modulation",   int_t, 0, "Frequence modulation", 0, 0, 5, edit_method=frequence_modulation_enum)
gen.add("channel_modulation",     int_t, 0, "Chanel modulation", 0, 0, 15, edit_method=channel_enum)
gen.add("median_filter",         bool_t, 0, "enable/disable median filter",  False)
gen.add("average_filter",        bool_t, 0, "enable/disable average filter",  False)
gen.add("temporal_filter_factor", double_t,    0, "Temporal filter factor [0...1]",  0.0, 0, 1.0)
gen.add("temporal_filter_threshold", int_t,    0, "Temporal filter threshold [mm]",  0, 0, 10000)
gen.add("edge_filter_threshold",     int_t,    0, "Edge filter threshold [mm]",  0, 0, 10000)
gen.add("interference_detection_limit", int_t, 0, "Interference detection limit  (IDL) [LSB]",  0, 0, 10000)
gen.add("use_last_value", bool_t, 0,  "enable/disable use interference detection last value",  False)
gen.add("min_distance",   int_t,  0,  "Minimal distance [mm]",  0, 0, 50000)
gen.add("max_distance",   int_t,  0,  "Maximal distance [mm]",  6000, 0, 50000)

gen.add("roi_left_x",          int_t, 0,  "ROI left X",     0, 0,  312)
gen.add("roi_right_x",         int_t, 0,  "ROI right X",  319, 7,  319)
gen.add("roi_height",          int_t, 0,  "ROI height",   240, 8,  240)

gen.add("cartesian",           bool_t, 0,  "Enable/disbale cartesian transformation",  True)
gen.add("publish_temperature", bool_t, 0,  "Publish temperature data",  True)
gen.add("publish_images",      bool_t, 0,  "Publish image data",  True)
gen.add("publish_point_cloud", bool_t, 0,  "Publish point cloud data",  True)

exit(gen.generate(PACKAGE, "tof_cam660_node", "espros_cam660"))
